### ISSUE TITLE
bytes (and strings?) should support parallel iteration (zippered and non-)

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#19112]
### Summary of Problem

While working on AoC, I found that our `bytes` type doesn't support parallel iteration.  Unless I'm missing something, it seems that this is a no-brainer feature that one would want and that should be trivial to support.  See the linked future below for samples.

At first, I thought "I guess it makes sense that strings don't support this", but on reflection, I think they probably should too, even though it will be more expensive to compute in its naive form.  Here, I'm proposing we start with bytes and then fork off a new issue for strings once they're working (if we don't just tackle both simultaneously).

### Steps to Reproduce

**Associated Future Test(s):**
[`test/studies/adventOfCode/2021/bradc/futures/day16-bytesParIter.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/studies/adventOfCode/2021/bradc/futures/day16-bytesParIter.chpl
          ) #19028 

### Configuration Information

- Output of `chpl --version`: `chpl version 1.26.0 pre-release (56bd177336)`

**Comment by bradcray** [comment]
@e-kayrakli : I almost just slapped a "easy/straightforward" / "good first issue" label on here, but wanted to check with you before doing so (where I imagine I'd do the mentoring).

**Comment by e-kayrakli** [comment]
Yeah, I don't have anything significant to add to your observations at a high level. Though looking at the code, here are some notes:

- There are `chpl_bytes` iterator for both types, that is used as `x.bytes()`.
- I am worried that `bytes.chpl_bytes` does not `localize` the `bytes`'s data, but string's does. I think both should.
- I think `bytes.items` should just forward to `bytes.chpl_bytes`.
- `bytes.these` does forward to `bytes.chpl_bytes`
- If we add parallel versions, we should add all of them.
- We can make `string.chpl_bytes` parallel, too. And I think we should.
- Then, we should add parallel `string.these` and `string.items` just to give an error for now.
  - As you alluded to there's some complications for string, but I don't think it would be noticeably expensive.
    - Nuts-and-bolts: The beginning byte of a codepoint is easily discernible, it has its MSB set (I need to check this information; it's been a while) So, you can just chop up the byte buffer of the string equally, and just walk back (or forward) just a few bytes until you see the beginning byte of a codepoint. So, it doesn't require a sequential O(n) pass to find the cutting points.

**Comment by e-kayrakli** [comment]
I guess I haven't really answered your question but just danced around it. You can make an easy/straightforward issue out of this, but there are some engineering that needs to be done to have the ideal solution. I am probably not as positive as you are about the easiness of it, but I'm ok with it. I can help shepherding the work.

For some reason, I find "easy" a better label than "good first issue". Where I see the latter as next to 0 implementation work, but a good way to get the hang of the contribution process.

**Comment by bradcray** [comment]
Yeah, if the only way to close this were to address all the things in your first comment, I agree that this isn't either easy or a good first issue.  It also makes me glad that I opened this issue rather than diving into the implementation.  But it doesn't make me any less sad that we don't have this ability today.
